app/gui2/src/util/ast/aliasAnalysis.ts

```diff
@@ -0,0 +1,276 @@
+  withNewScopeOver(nodeOrRange: ContentRange | Tree | Token, f: () => undefined) {
+    const range = parsedTreeOrTokenRange(nodeOrRange)
+    const scope = new Scope(range, this.scopes.top)
+    this.scopes.withPushed(scope, f)
+  }
+
+  /** Execute the given function while the given context is temporarily pushed to the stack. */
+  withContext(context: Context, f: () => void) {
+    return this.contexts.withPushed(context, f)
+  }
+
+  /** Marks given token as a binding, i.e. a definition of a new variable. */
+  bind(token: Token) {
+    const scope = this.scopes.top
+    const identifier = readTokenSpan(token, this.code)
+    const range = parsedTreeOrTokenRange(token)
+    log(`Binding ${identifier}@[${range}]`)
+    scope.bindings.set(identifier, token)
+    assert(!this.aliases.has(range), `Token at ${range} is already bound.`)
+    this.aliases.set(range, new ObjectKeyedSet())
+  }
+
+  addConnection(source: Token, target: Token) {
+    const sourceRange = parsedTreeOrTokenRange(source)
+    const targetRange = parsedTreeOrTokenRange(target)
+    const targets = this.aliases.get(sourceRange)
+    if (targets != null) {
+      log(
+        `Usage of ${readTokenSpan(
+          source,
+          this.code,
+        )}@[${targetRange}] resolved to [${sourceRange}]`,
+      )
+      targets.add(targetRange)
+    } else {
+      console.warn(`No targets found for source range ${sourceRange}`)
+    }
+  }
+
+  /** Marks given token as a usage, i.e. a reference to an existing variable. */
+  use(token: Token) {
+    const identifier = readTokenSpan(token, this.code)
+    const range = parsedTreeOrTokenRange(token)
+
+    for (const scope of this.scopes.valuesFromTop()) {
+      const binding = scope.bindings.get(identifier)
+      if (binding != null) {
+        this.addConnection(binding, token)
+        return
+      }
+    }
+
+    log(`Usage of ${identifier}@[${range}] is unresolved.`)
+    this.unresolvedSymbols.add(range)
+  }
+
+  /** Method that processes a single AST node. */
+  public process() {
+    this.processNode(this.ast)
+  }
+
+  processToken(token?: Token): void {
+    if (token == null) {
+      return
+    }
+
+    const repr = readTokenSpan(token, this.code)
+
+    // We require first character to be lowercase, to avoid working with type names.
+    if (repr.length > 0 && repr.charAt(0) !== repr.charAt(0).toUpperCase()) {
+      if (this.contexts.top === Context.Pattern) {
+        this.bind(token)
+      } else {
+        this.use(token)
+      }
+    }
+  }
+
+  /** Method that processes a single AST node. */
+  processNode(node?: Tree): void {
+    if (node == null) {
+      return
+    }
+
+    const range = parsedTreeOrTokenRange(node)
+    const repr = readAstOrTokenSpan(node, this.code)
+
+    log(
+      '\nprocessNode',
+      astPrettyPrintType(node),
+      '\n====\n',
+      readAstOrTokenSpan(node, this.code),
+      '\n====\n',
+    )
+
+    if (node.type === Tree.Type.BodyBlock) {
```

Comment:
This whole block should definitely be a `switch(node.type)` instead of long `else if` chain.

Reply:
Refactored.